pkg/cri/server/container_events.go

```diff
@@ -21,13 +21,49 @@ import (
 )
 
 func (c *criService) GetContainerEvents(r *runtime.GetEventsRequest, s runtime.RuntimeService_GetContainerEventsServer) error {
-	// TODO (https://github.com/containerd/containerd/issues/7318):
-	// replace with a real implementation that broadcasts containerEventsChan
-	// to all subscribers.
-	for event := range c.containerEventsChan {
-		if err := s.Send(&event); err != nil {
-			return err
+	done := make(chan error, 1)
+
+	c.containerEventsClients.Store(s, done)
+
+	err := <-done
+
+	c.containerEventsClients.Delete(s)
+
+	return err
+}
+
+func (c *criService) broadcastEvents() {
+	defer c.containerEventsClients.Range(func(_, value any) bool {
+		done, ok := value.(chan error)
+		if !ok {
+			return true
 		}
+
+		select {
+		case done <- nil:
+		default:
+		}
+
+		return true
+	})
+
+	for containerEvent := range c.containerEventsChan {
+		c.containerEventsClients.Range(func(key, value any) bool {
+			stream, ok := key.(runtime.RuntimeService_GetContainerEventsServer)
+			if !ok {
+				return true
+			}
+
+			done, ok := value.(chan error)
+			if !ok {
+				return true
+			}
+
+			if err := stream.Send(&containerEvent); err != nil {
+				done <- err
+			}
```

Comment:
Couldn't we:

```go
done <- stream.Send(&containerEvent)
return true
```

and then get rid of the defer at the top? I'd probably rename `done` to `errCh` also, I feel like I see done channels as just a "this thing finished, move on" and generally they just get sent empty structs down them.

Reply:
Doing
```go
done <- stream.Send(&containerEvent)
return true
```
will stop the broadcaster after the first send. See the Cirrus CI logs: https://cirrus-ci.com/task/6063714311340032

We want a continuous stream of events here.

`defer` is probably not bad up there as it allows us to gracefully stop sending events when closing `CRIService`.

Renaming `done` to `errCh` makes sense 👍

Reply:
Ack, I must've missed the logic reviewing on my phone 😳. I'm not a huge fan of the defer, so I'll try and take a look tomorrow to see what we could do. Thanks for the quick response!

Reply:
Hi @dcantah,
I hope you're well — just a friendly nudge about the PR. I'm looking forward to your feedback whenever you have a chance. Thanks!

Reply:
Hey! Sorry for delay. Will look shortly, thanks for your patience here

Reply:
Honestly, I can't think of much nicer without making this more complicated. Perhaps just handle the stream context finishing though

Reply:
```go
for containerEvent := range c.containerEventsChan {
	c.containerEventsClients.Range(func(key, value any) bool {
		stream, ok := key.(runtime.RuntimeService_GetContainerEventsServer)
		if !ok {
			return true
		}

		errCh, ok := value.(chan error)
		if !ok {
			return true
		}

		select {
		case <-stream.Context().Done():
			errCh <- stream.Context().Err()
		default:
			if err := stream.Send(&containerEvent); err != nil {
				errCh <- err
			}
		}

		return true
	})
}
```